File: mic_sims_files/array_geometry.py
"""
array_geometry.py

One place to define where the microphones physically are.

Every script that needs geometry imports it from here, so trying a new
array layout means editing ACTIVE below (or adding a LAYOUTS entry), not
hunting through several files. localization_sim.py deliberately keeps its
own 10 cm square MIC_POS and is NOT driven by this file: that square is
the validated reference the firmware port gets compared against, and it
should stay fixed.

Coordinates are metres, (x, y) per mic, in mic order 0..3.

The mic labelling matches the physical build: mic0 top-left, mic1
top-right, mic2 bottom-left, mic3 bottom-right, with +x pointing right and
+y pointing up.

    +y
     ^   mic0 .          . mic1     (top)
     |
     +---> +x
         mic2 .          . mic3     (bottom)

Bearings reported elsewhere are degrees counterclockwise from +x, so 0 deg
is off the right edge (toward mic1/mic3), 90 deg is off the top edge
(toward mic0/mic1), 180 deg is off the left edge.

This order is not cosmetic. The firmware always emits the four channels as
mic0, mic1, mic2, mic3, and every script assigns MIC_POS[m] to channel m,
so the row order here MUST match how the mics are physically wired: the mic
you place at top-left must be the one wired as channel 0 (PE7, SEL=GND, per
the mic map in CLAUDE.md). Get this wrong and bearings come out reflected
even though every delay is correct.

BUILD NOTE, and this one matters: along the LONG axis you can count holes,
because the 0.1 inch (2.54 mm) column pitch is exact and uninterrupted.
Across the WIDTH you cannot. The centre channel of each board, the power
rails, and the glue seam between the two boards all break the grid, so the
width is whatever your boards happen to give. MEASURE the width, do not
count it.

Either way the final numbers must come from calipers on the mic PORTS, not
the header pins: the port is offset from the pins on a breakout board. For
scale, 1 mm of position error is about 0.047 samples of delay, well under
a degree, so +/- 1 mm is good enough.
"""

import numpy as np

C_SOUND = 343.0
FS = 16000


def rect(width_x, height_y):
    """
    Four mics at the corners of a rectangle centred on the origin, in the
    physical labelling mic0 top-left, mic1 top-right, mic2 bottom-left,
    mic3 bottom-right (+x right, +y up).

    width_x  is the left-to-right span (mic0<->mic1 and mic2<->mic3).
    height_y is the top-to-bottom span (mic0<->mic2 and mic1<->mic3).
    """
    w, h = width_x / 2.0, height_y / 2.0
    return np.array([
        [-w,  h],   # mic0 top-left
        [ w,  h],   # mic1 top-right
        [-w, -h],   # mic2 bottom-left
        [ w, -h],   # mic3 bottom-right
    ])


# Named layouts. Add entries here when trying something new; the scoring
# script picks them all up automatically.
#
#   positions : metres
#   measured  : True only once the numbers came off calipers
#   note      : what it needs and what it is for
# Measured dimensions of the built array. Left-to-right span and
# top-to-bottom span, in metres. Edit these two numbers if you re-measure
# or re-space the array; everything downstream follows.
BUILT_WIDTH_X = 0.0925    # 9.25 cm, mic0<->mic1 (measured with calipers)
BUILT_HEIGHT_Y = 0.099    # 9.90 cm, mic0<->mic2 (measured with calipers)

LAYOUTS = {
    "9.25x9.9-measured": {
        "positions": rect(BUILT_WIDTH_X, BUILT_HEIGHT_Y),
        "measured": True,
        "note": "THE BUILT ARRAY. 9.25 cm wide x 9.9 cm tall, mics measured "
                "port to port with calipers. mic0 top-left, mic1 top-right, "
                "mic2 bottom-left, mic3 bottom-right. Near-square, no blind "
                "directions.",
    },
    "9.25x10-nominal": {
        "positions": rect(0.100, 0.0925),
        "measured": False,
        "note": "Build target. Two full-size breadboards glued. 10 cm along "
                "the long axis is 39-40 hole pitches; 9.25 cm width is the "
                "full span across both boards including rails. Most uniform "
                "option (condition number 1.08), no blind directions.",
    },
    "9.25x12-nominal": {
        "positions": rect(0.120, 0.0925),
        "measured": False,
        "note": "Stretch goal. Slightly better accuracy from the longer "
                "aperture, still no blind directions.",
    },
    "9.25x16-nominal": {
        "positions": rect(0.160, 0.0925),
        "measured": False,
        "note": "Full board length. Best simulated accuracy, but the widest "
                "aperture also carries the most spatial aliasing risk in a "
                "reverberant room, which the simulation does not model. Try "
                "it after the 10 cm build works.",
    },
    "2.8x10-single-board": {
        "positions": rect(0.100, 0.028),
        "measured": False,
        "note": "Superseded. What one breadboard alone allowed. Kept for "
                "comparison: four blind cones about 30 deg off the long "
                "axis where error reaches 8 deg.",
    },
    "10cm-square-reference": {
        "positions": rect(0.100, 0.100),
        "measured": False,
        "note": "The geometry localization_sim.py is validated against. "
                "Does not fit the boards; here for scoring comparison only.",
    },
}

# Which layout the analysis scripts should use.
ACTIVE = "9.25x9.9-measured"


def active_positions():
    return LAYOUTS[ACTIVE]["positions"]


def is_measured():
    return LAYOUTS[ACTIVE]["measured"]


def pairs(n=4):
    return [(i, j) for i in range(n) for j in range(i + 1, n)]


def describe(positions):
    """
    Geometric health of a layout.

    rank < 2 means the array is collinear and direction estimation cannot
    work: all baselines are parallel, so the across-axis component is
    unconstrained and the answer collapses to 0 or 180 regardless of
    truth. Condition number near 1 means accuracy is uniform in all
    directions; large means strongly direction dependent.
    """
    P = np.asarray(positions)
    A = np.array([P[i] - P[j] for (i, j) in pairs(len(P))])
    rank = int(np.linalg.matrix_rank(A))
    s = np.linalg.svd(A, compute_uv=False)
    cond = float(s[0] / s[1]) if len(s) > 1 and s[1] > 1e-12 else float("inf")
    ap = max(np.linalg.norm(P[i] - P[j]) for (i, j) in pairs(len(P)))
    return {
        "rank": rank,
        "cond": cond,
        "aperture_m": ap,
        "aperture_samples": ap / C_SOUND * FS,
        "span_x": float(P[:, 0].max() - P[:, 0].min()),
        "span_y": float(P[:, 1].max() - P[:, 1].min()),
    }


if __name__ == "__main__":
    print(f"ACTIVE layout: {ACTIVE}")
    print(f"  measured: {is_measured()}")
    print(f"  {LAYOUTS[ACTIVE]['note']}\n")
    for name, spec in LAYOUTS.items():
        d = describe(spec["positions"])
        mark = " <- ACTIVE" if name == ACTIVE else ""
        print(f"{name:<24} {d['span_x']*100:5.1f} x {d['span_y']*100:5.1f} cm  "
              f"aperture {d['aperture_m']*100:5.1f} cm "
              f"({d['aperture_samples']:4.1f} samples)  "
              f"rank {d['rank']}  cond {d['cond']:5.2f}{mark}")

File: mic_test/Core/Src/array_geometry.c
/* array_geometry.c - the layout table and geometric health checks.
 *
 * Firmware mirror of mic_sims_files/array_geometry.py. If you edit the
 * dimensions here, edit them there too (and vice versa), or the board and the
 * offline analysis will disagree about where the mics are while both look
 * perfectly self-consistent.
 */
#include "array_geometry.h"

#include <math.h>

/* Measured dimensions of the built array, metres. Left-to-right span and
   top-to-bottom span. These two numbers are the ones to change if you
   re-measure or re-space the current build; everything downstream follows.
   Same values as BUILT_WIDTH_X / BUILT_HEIGHT_Y in array_geometry.py. */
#define BUILT_WIDTH_X    0.0925f   /* 9.25 cm, mic0<->mic1, ruler */
#define BUILT_HEIGHT_Y   0.0990f   /* 9.90 cm, mic0<->mic2, ruler */

/* Corner positions of a rectangle centred on the origin, in the physical
   labelling mic0 top-left, mic1 top-right, mic2 bottom-left, mic3
   bottom-right. Written out rather than computed so the table stays a plain
   initialiser you can read at a glance. RECT_* mirrors rect() in the Python. */
#define RECT(w, h)  { { -(w) / 2.0f,  (h) / 2.0f },   /* mic0 top-left     */ \
                      {  (w) / 2.0f,  (h) / 2.0f },   /* mic1 top-right    */ \
                      { -(w) / 2.0f, -(h) / 2.0f },   /* mic2 bottom-left  */ \
                      {  (w) / 2.0f, -(h) / 2.0f } }  /* mic3 bottom-right */

/* Named layouts. Order MUST match the enum in array_geometry.h.
   The superseded 2.8x10 single-board rectangle from the Python registry is
   deliberately not carried over: it has four blind cones and there is no
   reason to be able to select it by accident on hardware. */
static const array_layout_t kLayouts[GEOM_LAYOUT_COUNT] = {
    [GEOM_9_25X9_9_MEASURED] = {
        .name     = "9.25x9.9-measured",
        .pos      = RECT(BUILT_WIDTH_X, BUILT_HEIGHT_Y),
        .measured = 1,
        .note     = "THE BUILT ARRAY, calipered port to port. Near-square, "
                    "no blind directions.",
    },
    [GEOM_9_25X10_NOMINAL] = {
        .name     = "9.25x10-nominal",
        .pos      = RECT(0.100f, 0.0925f),
        .measured = 0,
        .note     = "Build target, nominal numbers. Most uniform option.",
    },
    [GEOM_9_25X12_NOMINAL] = {
        .name     = "9.25x12-nominal",
        .pos      = RECT(0.120f, 0.0925f),
        .measured = 0,
        .note     = "Stretch goal. Longer aperture, still no blind cones.",
    },
    [GEOM_9_25X16_NOMINAL] = {
        .name     = "9.25x16-nominal",
        .pos      = RECT(0.160f, 0.0925f),
        .measured = 0,
        .note     = "Full board length. Best simulated accuracy, most "
                    "spatial aliasing risk in a real room.",
    },
    [GEOM_10CM_SQUARE_REFERENCE] = {
        .name     = "10cm-square-reference",
        .pos      = RECT(0.100f, 0.100f),
        .measured = 0,
        .note     = "The geometry localization_sim.py is validated against. "
                    "Select this to compare the port against the reference.",
    },
};

/* The six unique pairs, in the same order as the Python PAIRS list. */
static const int kPairs[LOC_NUM_PAIRS][2] = {
    { 0, 1 }, { 0, 2 }, { 0, 3 }, { 1, 2 }, { 1, 3 }, { 2, 3 },
};


const array_layout_t *Geom_Active(void)
{
    return &kLayouts[GEOM_ACTIVE];
}


const array_layout_t *Geom_Get(int index)
{
    if (index < 0 || index >= GEOM_LAYOUT_COUNT) return 0;
    return &kLayouts[index];
}


void Geom_Pair(int pair_index, int *mic_i, int *mic_j)
{
    if (pair_index < 0 || pair_index >= LOC_NUM_PAIRS) {
        *mic_i = 0;
        *mic_j = 0;
        return;
    }
    *mic_i = kPairs[pair_index][0];
    *mic_j = kPairs[pair_index][1];
}


float Geom_Spacing(const array_layout_t *layout, int mic_i, int mic_j)
{
    const float dx = layout->pos[mic_i][0] - layout->pos[mic_j][0];
    const float dy = layout->pos[mic_i][1] - layout->pos[mic_j][1];
    return sqrtf(dx * dx + dy * dy);
}


float Geom_MaxTauSamples(const array_layout_t *layout, int mic_i, int mic_j)
{
    return Geom_Spacing(layout, mic_i, mic_j) / LOC_METRES_PER_SAMPLE;
}


/*
 * Geometric health, the C counterpart of describe() in array_geometry.py.
 *
 * Build the same 6x2 baseline matrix the least squares fit solves, where row
 * p is (pos_i - pos_j) for pair p, and ask two questions about it:
 *
 *   rank  - must be 2. A collinear array is rank 1: every baseline is
 *           parallel, the across-axis component of the direction vector is
 *           unconstrained, and the reported angle collapses to 0 or 180
 *           regardless of truth. Localize_Run refuses to answer in that case.
 *   cond  - ratio of the singular values. Near 1 means accuracy is uniform in
 *           every direction. Large means narrow bearings where error blows up.
 *
 * numpy gets the singular values from an SVD. Here the matrix is only 6x2, so
 * the singular values of A are the square roots of the eigenvalues of the 2x2
 * symmetric matrix A'A, which has a closed form. Same answer, no iteration,
 * about twenty lines.
 */
void Geom_Describe(const array_layout_t *layout, geom_info_t *out)
{
    float sxx = 0.0f, sxy = 0.0f, syy = 0.0f;   /* entries of A'A */

    for (int p = 0; p < LOC_NUM_PAIRS; p++) {
        const int i = kPairs[p][0];
        const int j = kPairs[p][1];
        const float bx = layout->pos[i][0] - layout->pos[j][0];
        const float by = layout->pos[i][1] - layout->pos[j][1];
        sxx += bx * bx;
        sxy += bx * by;
        syy += by * by;
    }

    /* Eigenvalues of [[sxx, sxy], [sxy, syy]]. Both are real and >= 0 because
       A'A is symmetric positive semi-definite. */
    const float trace = sxx + syy;
    const float diff  = sxx - syy;
    const float root  = sqrtf(diff * diff + 4.0f * sxy * sxy);
    float lam_hi = 0.5f * (trace + root);
    float lam_lo = 0.5f * (trace - root);
    if (lam_lo < 0.0f) lam_lo = 0.0f;           /* clamp rounding noise */

    const float s_hi = sqrtf(lam_hi);
    const float s_lo = sqrtf(lam_lo);

    /* Same relative tolerance idea numpy's matrix_rank uses: a singular value
       counts only if it is not negligible next to the largest one. */
    const float tol = 1e-6f * s_hi;
    out->rank = 0;
    if (s_hi > tol) out->rank++;
    if (s_lo > tol) out->rank++;

    out->cond = (s_lo > 1e-12f) ? (s_hi / s_lo) : INFINITY;

    /* Aperture: the widest mic-to-mic distance, and what that is worth in
       samples of delay. This is the largest lag any pair can legitimately
       produce, which is what bounds the GCC-PHAT search. */
    float widest = 0.0f;
    for (int p = 0; p < LOC_NUM_PAIRS; p++) {
        const float d = Geom_Spacing(layout, kPairs[p][0], kPairs[p][1]);
        if (d > widest) widest = d;
    }
    out->aperture_m       = widest;
    out->aperture_samples = widest / LOC_METRES_PER_SAMPLE;

    float xmin = layout->pos[0][0], xmax = layout->pos[0][0];
    float ymin = layout->pos[0][1], ymax = layout->pos[0][1];
    for (int m = 1; m < LOC_NUM_MICS; m++) {
        if (layout->pos[m][0] < xmin) xmin = layout->pos[m][0];
        if (layout->pos[m][0] > xmax) xmax = layout->pos[m][0];
        if (layout->pos[m][1] < ymin) ymin = layout->pos[m][1];
        if (layout->pos[m][1] > ymax) ymax = layout->pos[m][1];
    }
    out->span_x = xmax - xmin;
    out->span_y = ymax - ymin;
}
